src/cargo/core/workspace.rs

```diff
@@ -578,25 +581,39 @@ fn expand_member_path(path: &Path) -> CargoResult<Vec<PathBuf>> {
 }
 
 fn is_excluded(members: &Option<Vec<String>>,
-               exclude: &[String],
+               exclude: &Gitignore,
                root_path: &Path,
                manifest_path: &Path) -> bool {
-    let excluded = exclude.iter().any(|ex| {
-        manifest_path.starts_with(root_path.join(ex))
-    });
+    let exclude_match = exclude.matched_path_or_any_parents(manifest_path, false);
 
     let explicit_member = match *members {
         Some(ref members) => {
-            members.iter().any(|mem| {
-                manifest_path.starts_with(root_path.join(mem))
-            })
+            members.iter()
+                .filter(|p| manifest_path.starts_with(root_path.join(p)))
```

Comment:
I think that members can contain globs. That is, `p` might be like `foo/*` here, and so this `starts_with` will fail.